feat: circuit based automata with support for predicates

bollu opened the pull request:
We add a `Predicate` along with a `Term` which models bitvector predicates.

- [x] implement BV predicates
- [x] reduce to `decideAllZeroes`
- [x] implement reflection for predicates (`eq`, `neq`).
- [x] implement new meta-code to convert to negation normal form.
- [x] implement coverage for the other predicates (`<u`, `<=u`, `<s`, `<=s`)

bollu committed:
chore: implement scanAnd and scanOr

To decide equality of bitvectors, we check that the stream
is all zeroes.

We want to widen the interpretation of predicates into automata
whose output is *eventually* always zero.

We will show that `isAlwaysZero x <-> isEventuallAlwaysZero (scan x)`,
which allows us to translate equality into the predicate based
framework.

bollu committed:
chore: add proofs around scan

bollu committed:
chore: make progress

bollu committed:
chore: remove sorry from decide

bollu committed:
feat: add predicate evaluation for all predicates

bollu committed:
feat: add predicate proof

bollu committed:
feat: start adding reflection implementation

bollu committed:
chore: reflect into bitvec

bollu committed:
feat: add reflection metaprogramming

tobiasgrosser commented:
Wow, any examples of inequalities already in the test cases?

bollu commented:
No, because that needs meta-code to hook it up, but I'll do it over the next day or two :)

tobiasgrosser commented:
OK.

bollu committed:
chore: add NNF pass

bollu committed:
chore: write reflection

tobiasgrosser commented on SSA/Experimental/Bits/Fast/Reflect.lean:
```diff
@@ -130,57 +139,197 @@ TODO(@bollu): We also assume that the goals are in negation normal form, and if
+  sorry
+
+/-- error: expcted width to be a free variable, found '10' (bitwidth with width '10' is 'b') -/
+#guard_msgs in example : ∀ (a b : BitVec 10), a = b := by
```
```suggestion
/-- error: expected width to be a free variable, found '10' (bitwidth with width '10' is 'b') -/
#guard_msgs in example : ∀ (a b : BitVec 10), a = b := by
```

bollu replied:
@tobiasgrosser Thanks for the patch. Unfortunately, I cannot accept the suggestion, since it will break the build: For these changes, it would help a bit more if the patch could be made against the part of the codebase that emits this error. Thanks for pointing the typo out, though, I appreciate it!

tobiasgrosser replied:
I don't get this. I just fixed the typo? Why does this break things?

bollu committed:
chore: fixup decide by bringing back sorrys :)

bollu committed:
chore: make comment

bollu committed:
chore: reflow

bollu committed:
chore: try to use the Decidable instance

bollu committed:
chore: stash work on ult/slt

bollu committed:
Revert "chore: stash work on ult/slt"

This reverts commit f6be8b59c401603e16a7834b2c0c074c11aff537.

bollu committed:
chore: add reflect impl

bollu committed:
chore: use native_decide

bollu committed:
chore: cleanup dead code

bollu committed:
chore: nuke old tactic frontend, which uses the old definition of Nat -> BitStream that is way harder to reflect and get defeqs than (List Bitstream)

bollu committed:
chore: eliminate sorrys

bollu committed:
chore: finish building FSMs for ule, ult, sle, slt

bollu committed:
chore: replace Tactic file with Reflect file

The tactic implementation was wonky IMO,
so I replaced it with the 'correct' style of proof by reflection.

tobiasgrosser commented on SSA/Experimental/Bits/Fast/Tactic.lean:
```diff
@@ -1,491 +0,0 @@
--/
-
-def alive_1 {w : ℕ} (x x_1 x_2 : BitVec w) : (x_2 &&& x_1 ^^^ x_1) + 1#w + x = x - (x_2 ||| ~~~x_1) := by
-  bv_automata
```
These test cases might be worth reviving whenever your tactic is powerful enough.

bollu replied:
Done, I've revived them now.

bollu committed:
chore: run NNF and negate full goal with hyps

bollu committed:
chore: support the full predicate fragment in reflection, now think about why we don't decide right!

bollu committed:
chore: reflect the full term language

bollu committed:
chore: also reflect signed and unsigned compares

bollu commented on SSA/Experimental/Bits/Fast/Reflect.lean:
```diff
@@ -0,0 +1,671 @@
+
+open NNF in
+
+example (w : Nat) (a b : BitVec w) : a + b = b + a := by
```
@tobiasgrosser I've left examples here, and I'm still adding to them at the moment. Feel invited to add new ones, which will help me flag down what isn't working right now!

bollu committed:
chore: add more examples of slt, sle, also record thoughts on shifts

bollu committed:
chore: more examples

bollu committed:
chore: jiggle with automata

bollu committed:
chore: fixup not equals, we need a scanAnd

bollu committed:
feat: Fixup all predicates with ult, slt.

This fixes the automata implementations.

There seems to be a metaprogramming bug in the slt predicate
which trips two examples up. Unsure what, will debug next!

Anyway, this shows that the core algorithmics is sound.
Next step: prove the algorithm correct.

bollu committed:
fix: Bugfix in reflection term

Was incorrectly denoting slt to sle and sle to slt, fixed!
(Note that this does not effect `eval`, which is a separate construct,
and what is actually run by the decision procedures).

bollu committed:
chore: steal all examples from Tactic.lean

The examples, when generalized to arbitrary bitwidth, work,
save for one example that needs us to have `2#w`, which I don't support yet.
We need to write a reflector into `Term.const`, and the associated FSM construction.
Will do this next.

bollu commented:
Core support is now complete. We can add the following incrementally:
- [ ] Constants other than `0`, `1`, `-1` by supporting arbitrary `BitVec.ofNat w n` declarations
- [ ] `shl bv k` where `k` is a compile-time known constant.
- [ ] `lshr bv k` where `k` is a compile-time known constant.
- I *don't think* that `ashr bv k` is possible, since the arithmetic shifting is an "irregular" operation across different bitwidths. I wonder if reading the p-adic decision procedure will give me some insights here.

bollu commented:
The next step seems to be to actually verify the new algorithm.

tobiasgrosser reviewed (approved):
Let's merge, and then iterate.
